syntax = "proto3";

import public "geyser.proto";

package fumarole_v2;

service Fumarole {
  rpc GetConsumerGroupInfo(GetConsumerGroupInfoRequest) returns (ConsumerGroupInfo) {}
  rpc ListConsumerGroups(ListConsumerGroupsRequest) returns (ListConsumerGroupsResponse) {}
  rpc DeleteConsumerGroup(DeleteConsumerGroupRequest) returns (DeleteConsumerGroupResponse) {}
  rpc CreateConsumerGroup(CreateConsumerGroupRequest) returns (CreateConsumerGroupResponse) {}

  rpc DownloadBlock(DownloadBlockShard) returns (stream DataResponse) {}

  // Represents subscription to the data plane
  rpc SubscribeData(stream DataCommand) returns (stream DataResponse) {}

  // Represents subscription to the control plane
  rpc Subscribe(stream ControlCommand) returns (stream ControlResponse) {}
}

message GetConsumerGroupInfoRequest {
  string consumer_group_name = 1;
}

message DeleteConsumerGroupRequest {
  string consumer_group_name = 1;
}

message DeleteConsumerGroupResponse {
  bool success = 1;
}

message ListConsumerGroupsRequest {}

message ListConsumerGroupsResponse {
  repeated ConsumerGroupInfo consumer_groups = 1;
}


message ConsumerGroupInfo {
  string id = 1;
  string consumer_group_name = 2;
  bool is_stale = 3;
}

message GetSlotLagInfoRequest {
  string consumer_group_name = 1;
}

message BlockFilters {
    map<string, geyser.SubscribeRequestFilterAccounts> accounts = 1;
    map<string, geyser.SubscribeRequestFilterTransactions> transactions = 2;
    map<string, geyser.SubscribeRequestFilterEntry> entries = 3;
    map<string, geyser.SubscribeRequestFilterBlocksMeta> blocks_meta = 4;
}

message DownloadBlockShard {
    bytes blockchain_id = 1;
    bytes block_uid = 2;
    int32 shard_idx = 3;
    optional BlockFilters blockFilters = 4;
}


message Ping {
  uint32 ping_id = 1;
}

message Pong {
  uint32 ping_id = 1;
}

message DataCommand {
  oneof command {
    DownloadBlockShard download_block_shard = 1;
    Ping ping = 2;
  } 
}

message BlockShardDownloadFinish {}

message BlockNotFound {
  bytes blockchain_id = 1;
  bytes block_uid = 2;
  int32 shard_idx = 3;
}

message DataError {
  oneof error {
    BlockNotFound not_found = 1;
  }
}

message DataResponse {
  oneof response {
    geyser.SubscribeUpdate update = 1;
    BlockShardDownloadFinish block_shard_download_finish = 2;
  }
}

message CommitOffset {
  int64 offset = 1;
  int32 shard_id = 2;
}


message PollBlockchainHistory {
  int32 shard_id = 1;
  optional int64 from = 2;
  optional int64 limit = 3;
}

message BlockchainEvent {
  int64 offset = 1;
  bytes blockchain_id = 2;
  bytes block_uid = 3;
  uint32 num_shards = 4;
  uint64 slot = 5;
  optional uint64 parent_slot = 6;
  geyser.CommitmentLevel commitment_level = 7;
  int32 blockchain_shard_id = 8;
  optional string dead_error = 9;
}

message BlockchainHistory {
  repeated BlockchainEvent events = 1;
}

message JoinControlPlane {
  optional string consumer_group_name = 1;
}

message ControlCommand {
    oneof command {
        JoinControlPlane initial_join = 1;
        CommitOffset commit_offset = 2;
        PollBlockchainHistory poll_hist = 3;
        Ping ping = 4;
    }
}

message ControlResponse {
    oneof response {
        InitialConsumerGroupState init = 1;
        CommitOffsetResult commit_offset = 2;
        BlockchainHistory poll_hist = 3;
        Pong pong = 4;
    }
}


message CommitOffsetResult {
  int64 offset = 1;
  int32 shard_id = 2;
}

message InitialConsumerGroupState {
  bytes blockchain_id = 1;
  map<int32, int64> last_committed_offsets = 2; 
}


message CreateConsumerGroupResponse {
  string consumer_group_id = 1;
}

enum InitialOffsetPolicy {
  LATEST = 0;
  //FROM_SLOT = 1;
}

message CreateConsumerGroupRequest {
  string consumer_group_name = 1;
  InitialOffsetPolicy initial_offset_policy = 2;
  //optional uint64 from_slot = 3;
}